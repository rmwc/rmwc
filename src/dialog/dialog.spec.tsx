import React from 'react';
import { render, screen, waitFor } from '@testing-library/react';
import userEvent from '@testing-library/user-event';
import {
  Dialog,
  DialogTitle,
  DialogContent,
  DialogActions,
  DialogButton,
  SimpleDialog,
  createDialogQueue,
  DialogQueue
} from './';

describe('Dialog', () => {
  it('simple Dialog renders', () => {
    const { asFragment } = render(
      <SimpleDialog
        title="This is a simple dialog"
        header="Foo"
        body="You can pass the body prop, or anything you want as children."
        open
        onClose={(evt) => {}}
      />
    );

    expect(asFragment()).toMatchSnapshot();
  });

  it('simple Dialog renders with no children', () => {
    const { asFragment } = render(
      <SimpleDialog
        title="This is a simple dialog"
        open
        onClose={(evt) => {}}
      />
    );
    expect(asFragment()).toMatchSnapshot();
  });

  it('simple Dialog renders with children', () => {
    const { asFragment } = render(
      <SimpleDialog title="This is a simple dialog" open onClose={(evt) => {}}>
        Hello
      </SimpleDialog>
    );
    expect(asFragment()).toMatchSnapshot();
  });

  it('standard Dialog renders', () => {
    const el = render(
      <Dialog open onClose={(evt) => {}}>
        <DialogTitle>Dialog Title</DialogTitle>

        <DialogContent>This is a custom dialog.</DialogContent>
        <DialogActions>
          <DialogButton action="close">Cancel</DialogButton>
          <DialogButton action="accept">Sweet!</DialogButton>
        </DialogActions>
      </Dialog>
    );

    el.unmount();
  });
});

describe('DialogQueue', () => {
  it('renders', () => {
    const queue = createDialogQueue();
    const el = render(<DialogQueue dialogs={queue.dialogs} />);
    el.unmount();
  });

  it('alerts and accepts', async () => {
    const queue = createDialogQueue();
    render(<DialogQueue dialogs={queue.dialogs} />);

    queue.alert({ title: 'myAlert' });

    expect(screen.getByText('myAlert')).toBeInTheDocument();

    userEvent.click(screen.getByText('OK'));

    await waitFor(() =>
      expect(screen.queryByText('OK')).not.toBeInTheDocument()
    );
  });

  it('confirms and returns true', async () => {
    const queue = createDialogQueue();
    render(<DialogQueue dialogs={queue.dialogs} />);
    queue.confirm({ title: 'myConfirm' });

    expect(screen.getByText('myConfirm')).toBeInTheDocument();

    userEvent.click(screen.getByText('OK'));

    await waitFor(() =>
      expect(screen.queryByText('OK')).not.toBeInTheDocument()
    );
  });

  it('confirms and returns false', async () => {
    const queue = createDialogQueue();
    render(<DialogQueue dialogs={queue.dialogs} />);
    queue.confirm({ title: 'myConfirm' });

    expect(screen.getByText('myConfirm')).toBeInTheDocument();

    userEvent.click(screen.getByText('Cancel'));

    await waitFor(() =>
      expect(screen.queryByText('Cancel')).not.toBeInTheDocument()
    );
  });

  it('prompts and returns value', async () => {
    const queue = createDialogQueue();
    render(<DialogQueue dialogs={queue.dialogs} />);

    queue.prompt({ title: 'myPrompt' });

    expect(screen.getByText('myPrompt')).toBeInTheDocument();

    userEvent.click(screen.getByText('Accept'));

    await waitFor(() =>
      expect(screen.queryByText('Accept')).not.toBeInTheDocument()
    );
  });

  it('prompts and returns null', async () => {
    const queue = createDialogQueue();
    render(<DialogQueue dialogs={queue.dialogs} />);

    queue.prompt({ title: 'myPrompt', body: 'CUSTOM BODY' });

    expect(screen.getByText('myPrompt')).toBeInTheDocument();
    expect(screen.getByText('CUSTOM BODY')).toBeInTheDocument();

    userEvent.click(screen.getByText('Cancel'));

    await waitFor(() =>
      expect(screen.queryByText('Cancel')).not.toBeInTheDocument()
    );
  });

  it('supports onClose', async () => {
    const queue = createDialogQueue();
    render(<DialogQueue dialogs={queue.dialogs} />);
    const onClose = jest.fn();
    queue.alert({ onClose });

    userEvent.click(screen.getByText('OK'));

    await waitFor(() => expect(onClose).toHaveBeenCalled());
  });
});
