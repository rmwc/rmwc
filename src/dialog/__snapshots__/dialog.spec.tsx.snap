// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Dialog simple Dialog renders 1`] = `
<DocumentFragment>
  <div
    class="mdc-dialog mdc-dialog--opening"
  >
    <div
      class="mdc-dialog__container"
    >
      <div
        aria-modal="true"
        class="mdc-dialog__surface"
        role="alertdialog"
      >
        <h2
          class="mdc-dialog__title"
        >
          This is a simple dialogFoo
        </h2>
        <div
          class="mdc-dialog__content"
        >
          You can pass the body prop, or anything you want as children.
        </div>
        <div
          class="mdc-dialog__actions"
        >
          <button
            class="mdc-dialog__button mdc-button"
            data-mdc-dialog-action="close"
          >
            <div
              class="mdc-button__ripple"
            />
            <span
              class="mdc-button__label"
            >
              Cancel
            </span>
          </button>
          <button
            class="mdc-dialog__button mdc-button"
            data-mdc-dialog-action="accept"
            data-mdc-dialog-button-default="true"
          >
            <div
              class="mdc-button__ripple"
            />
            <span
              class="mdc-button__label"
            >
              Accept
            </span>
          </button>
        </div>
      </div>
    </div>
    <div
      class="mdc-dialog__scrim"
    />
  </div>
</DocumentFragment>
`;

exports[`Dialog simple Dialog renders with children 1`] = `
<DocumentFragment>
  <div
    class="mdc-dialog mdc-dialog--opening"
  >
    <div
      class="mdc-dialog__container"
    >
      <div
        aria-modal="true"
        class="mdc-dialog__surface"
        role="alertdialog"
      >
        <h2
          class="mdc-dialog__title"
        >
          This is a simple dialog
        </h2>
        <div
          class="mdc-dialog__content"
        >
          Hello
        </div>
        <div
          class="mdc-dialog__actions"
        >
          <button
            class="mdc-dialog__button mdc-button"
            data-mdc-dialog-action="close"
          >
            <div
              class="mdc-button__ripple"
            />
            <span
              class="mdc-button__label"
            >
              Cancel
            </span>
          </button>
          <button
            class="mdc-dialog__button mdc-button"
            data-mdc-dialog-action="accept"
            data-mdc-dialog-button-default="true"
          >
            <div
              class="mdc-button__ripple"
            />
            <span
              class="mdc-button__label"
            >
              Accept
            </span>
          </button>
        </div>
      </div>
    </div>
    <div
      class="mdc-dialog__scrim"
    />
  </div>
</DocumentFragment>
`;

exports[`Dialog simple Dialog renders with no children 1`] = `
<DocumentFragment>
  <div
    class="mdc-dialog mdc-dialog--opening"
  >
    <div
      class="mdc-dialog__container"
    >
      <div
        aria-modal="true"
        class="mdc-dialog__surface"
        role="alertdialog"
      >
        <h2
          class="mdc-dialog__title"
        >
          This is a simple dialog
        </h2>
        <div
          class="mdc-dialog__actions"
        >
          <button
            class="mdc-dialog__button mdc-button"
            data-mdc-dialog-action="close"
          >
            <div
              class="mdc-button__ripple"
            />
            <span
              class="mdc-button__label"
            >
              Cancel
            </span>
          </button>
          <button
            class="mdc-dialog__button mdc-button"
            data-mdc-dialog-action="accept"
            data-mdc-dialog-button-default="true"
          >
            <div
              class="mdc-button__ripple"
            />
            <span
              class="mdc-button__label"
            >
              Accept
            </span>
          </button>
        </div>
      </div>
    </div>
    <div
      class="mdc-dialog__scrim"
    />
  </div>
</DocumentFragment>
`;
